x-pack/platform/packages/shared/alerting-rule-utils/src/group_by_object_utils.ts

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { set } from '@kbn/safer-lodash-set';
+import { Group, FieldsObject } from './types';
+
+export const unflattenObject = <T extends object = FieldsObject>(object: object): T =>
+  Object.entries(object).reduce((acc, [key, value]) => {
+    set(acc, key, value);
+    return acc;
+  }, {} as T);
+
+export const flattenObject = (obj: FieldsObject, prefix: string = ''): FieldsObject =>
+  Object.keys(obj).reduce<FieldsObject>((acc, key) => {
+    const nextValue = obj[key];
+
+    if (nextValue) {
+      if (typeof nextValue === 'object' && !Array.isArray(nextValue)) {
+        const dotSuffix = '.';
+        if (Object.keys(nextValue).length > 0) {
+          return {
+            ...acc,
+            ...flattenObject(nextValue, `${prefix}${key}${dotSuffix}`),
+          };
+        }
+      }
+
+      const fullPath = `${prefix}${key}`;
+      acc[fullPath] = nextValue;
+    }
+
+    return acc;
+  }, {});
+
+export const getGroupByObject = (
+  groupBy: string | string[] | undefined,
+  groupValueSet: Set<string>
+): Record<string, object> => {
+  const groupKeyValueMappingsObject: Record<string, object> = {};
+  if (groupBy) {
+    groupValueSet.forEach((groupValueStr) => {
+      const groupValueArray = groupValueStr.split(',');
+      groupKeyValueMappingsObject[groupValueStr] = unflattenObject(
+        Array.isArray(groupBy)
+          ? groupBy.reduce((result, groupKey, index) => {
+              return { ...result, [groupKey]: groupValueArray[index]?.trim() };
+            }, {})
+          : { [groupBy]: groupValueStr }
+      );
+    });
+  }
+  return groupKeyValueMappingsObject;
+};
+
+export const getFormattedGroupBy = (
+  groupBy: string | string[] | undefined,
+  groupSet: Set<string>
+): Record<string, Group[]> => {
+  const groupByKeysObjectMapping: Record<string, Group[]> = {};
+  if (groupBy) {
+    groupSet.forEach((group) => {
+      const groupSetKeys = group.split(',');
```

Comment:
This just doesn't feel safe, if the value could contain a `,`.  It's possible to use either rule state, or perhaps alert state, to store the actual values so we don't have to parse from the instance id.  Feels like we should be doing that.

Using a `join(..., ',')` to create an instance id is kinda problematic as already mentioned, but having to parse that introduces additional problems.

Reply:
I have removed parsing of instance id and now I am using alert state to store grouping object which will be used to populate `context.grouping` for recovered alerts (commit: 0d5ece140223bc326c3ed4be9999d32b6cb23296).

Do I need to do something additional while adding a new field in alert state? For example, I was wandering when upgrading Kibana, would users face any issue if this field was not there in state before? I am handling undefined state but just want to understand how it works in general and are there any best practices to follow.